### PR TITLE
feat(HMS-3522): implement rbac middleware

### PR DESCRIPTION
**Pull request by avisiedo**
This change implements the rbac middleware to authorize the API
operations. It adds a generated client to communicate with
the rbac micro-service and a wrapper to enhance the developer
experience by abstracting the internals to retrieve and check
the permissions.

It adds primitives and data structures to map the rbac permissions,
and provide a loader so future changes on the API will require
update the rbac.yaml file and no code will change.

It adds unit tests to cover all the different components.

Things to enhance if necessary:
- The assignation of permissions is 1-1 per route.
- The prefix does not support multiple ones: This will be
  required to support /api/idmsvc/v1 and /api/idmsvc/v1.1
  prefixes (with only the major version and major and minor version).
- Use redis cache to enhance response of the service (see hmscontent-services).

Credits on RHEnvision and hmscontent-services repositories,
which inspired some of the changes on this PR and some of the
enhances for the future.

TODO:

- [x] More Unit tests.
- [x] Reorg commits.

> Once this change is merged, this could require some update on your current
> `configs/config.yaml` file for your local environment; please compare with the
> change into the template at `configs/config.example.yaml` file.

### CODE REVIEW

File: internal/infrastructure/middleware/rbac-data/permission.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+		return echo.NewHTTPError(http.StatusInternalServerError, "route is an empty string")
+	}
+	if method == "" {
+		return echo.NewHTTPError(http.StatusInternalServerError, "method is an empty string")
+	}
+	return nil
+}
+
+// GetPermission
+func (r *RBACMap) GetPermission(p, ro, m string) (RBACPermission, error) {
```
[the remark this answers is not in the document]
Comment (avisiedo) [reply internal/infrastructure/middleware/rbac-data/permission.go:141]: Thanks, it needs some refactor; I will do.

File: internal/infrastructure/middleware/rbac-data/permission.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	var (
+		err        error
+		methods    map[Method]RBACPermission
+		permission RBACPermission
+		ok         bool
+	)
+	if err = r.getPermissionGuards(p, ro, m); err != nil {
```
[the remark this answers is not in the document]
Comment (avisiedo) [reply internal/infrastructure/middleware/rbac-data/permission.go:148]: For keeping the cyclomatic complexity low; guards use to be a sequence of _if_ statements and each one increase this indicator; it keeps the function smaller and cleaner. Furthermore it keeps the unit tests simpler (I will check all the guards in one unit test; later just check one path when it fails in the caller function).

File: internal/infrastructure/middleware/rbac-data/permission.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+		return "", err
+	}
+	// FIXME HMS-3522 It can receive /api/idmsvc/v1 or /api/idmsvc/v1.0
+	ro = strings.TrimPrefix(ro, p)
+	if methods, ok = (*r)[Route(ro)]; !ok {
+		return "", echo.NewHTTPError(http.StatusUnauthorized, "Route not authorized")
```
[the remark this answers is not in the document]
Comment (avisiedo) [reply internal/infrastructure/middleware/rbac-data/permission.go:154]: You are totally right, at this stage the route path is defined and matched.

The intention here is to protect the developer (I) forget to update the _rbac_ mapping when a new public route is defined. Probably the error message is more accurate with "Route not mapped". I will update that; the same for the method.

File: internal/infrastructure/router/public.go [hunk 4]
```diff
@@ -54,6 +57,38 @@ func getOpenapiPaths(c RouterConfig) func() []string {
 	}
 }
 
+// FIXME-3522 The loader routing is not working
```
[the remark this answers is not in the document]
Comment (avisiedo) [reply internal/infrastructure/router/public.go:60]: The intention was to externalize in a `rbac.yaml` file the information, so future changes apply the same behavior, but only change the configuration at the `rbac.yaml` file. The loader is not full implemented, and probably will be delayed in a new ticket.

Load from a string with the yaml representation is a previous step to externalize; we can even embed the content of rbac.yaml into the generated binary (for instance: https://github.com/content-services/content-sources-backend/blob/main/pkg/external_repos/redhat_repos.go#L16).

File: internal/infrastructure/middleware/rbac-data/permission.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (frasertweedale) [reply internal/infrastructure/middleware/rbac-data/permission.go:154]: OK, it's not a showstopper for me.  Resolving the thread :)

File: internal/infrastructure/router/public.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (frasertweedale) [reply internal/infrastructure/router/public.go:60]: OK, no worries.  I just don't want to expend effort on it if it is not essential.  We can always make a tech-debt ticket to improve it later.

File: internal/infrastructure/middleware/rbac-data/permission.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (frasertweedale) [reply internal/infrastructure/middleware/rbac-data/permission.go]: I see it's resolved now, thanks!

**Review by frasertweedale — APPROVED** [review]
LGTM.  Thanks for tidying the commit history.  Merging now, then let's test in Stage :)

**Pull request merged**